src/renderer/rendered_window.rs

```diff
@@ -225,26 +220,80 @@ impl RenderedWindow {
+                self.scroll_v += acc * timestep;
+                self.current_scroll += self.scroll_v * timestep;
+                dt -= timestep;
+            }
+
+            if (self.current_scroll - scroll_destination).abs() < 0.01 {
+                self.reset_scroll();
             } else {
                 animating = true;
-                self.scroll_t = (self.scroll_t + dt / settings.scroll_animation_length).min(1.0);
             }
-
-            self.current_scroll = ease(
-                ease_out_expo,
-                self.start_scroll,
-                self.scroll_destination,
-                self.scroll_t,
-            );
         }
 
         animating
     }
 
+    fn draw_surface(&mut self, font_dimensions: Dimensions, default_background: Color) -> bool {
+        let image_size: (i32, i32) = (self.grid_size * font_dimensions).into();
+        let pixel_region = Rect::from_size(image_size);
+        let canvas = self.current_surface.surface.canvas();
+        canvas.clip_rect(pixel_region, None, Some(false));
+        canvas.clear(default_background);
+
+        let scroll_offset_lines = self.current_scroll.floor();
+        let scroll_offset = scroll_offset_lines - self.current_scroll;
+        let scroll_offset_pixels = (scroll_offset * font_dimensions.height as f32).round() as isize;
+        let mut has_transparency = false;
+
+        let mut background_paint = Paint::default();
+        background_paint.set_blend_mode(BlendMode::Src);
+        background_paint.set_alpha(default_background.a());
+
+        let lines: Vec<(Matrix, &Line)> = (0..self.grid_size.height as isize + 1)
+            .filter_map(|i| {
+                let line_index = (self.top_index + scroll_offset_lines as isize + i)
+                    .rem_euclid(self.lines.len() as isize)
+                    as usize;
+                if let Some(line) = &self.lines[line_index] {
+                    let mut m = Matrix::new_identity();
+                    m.set_translate((
+                        0.0,
+                        (scroll_offset_pixels + (i * font_dimensions.height as isize)) as f32,
+                    ));
+                    Some((m, line))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        for (matrix, line) in &lines {
+            if let Some(background_picture) = &line.background_picture {
+                has_transparency |= line.has_transparency;
+                canvas.draw_picture(background_picture, Some(matrix), Some(&background_paint));
+            }
+        }
+        let mut foreground_paint = Paint::default();
+        foreground_paint.set_blend_mode(BlendMode::SrcOver);
+        for (matrix, line) in &lines {
+            if let Some(foreground_picture) = &line.foreground_picture {
+                canvas.draw_picture(foreground_picture, Some(matrix), Some(&foreground_paint));
```

Comment:
/rant
I hate Skia, it's a big black box, and I have no idea what's going on. The approach I'm taking here is way too slow. 

With the code as it is, and my rendering changes in opengl on an rtx 3060 with a full screen 3840x1600 and a telescope window open, without blur and transparency it can render only around 90FPS. 

Commenting out this line, that is drawing only the background, so jsut a few hundred boxes gives just 160 FPS
Additionally commenting out the the background drawing gives 2000 FPS (that's more like it)

Note that I'm still rendereing from render target, which includes the `clear`  comand, so it shouldn't be some kind of strange synchronization issue between the reneder target and the backbuffer.

And another Skia rant, and upgrading to a newer version of Skia seems to force VSync on in some strange way incompatible with my other rendering changes.
/rant

This function probably does not need to run on every rendered frame. Only when the particular window has changed, which is every frame when scrolling. But at least it doesn't need to be run for the static windows.

But still something else has to be done, even if the FPS is ok for me (still can't render at my display refresh rate of 144 though), it probably won't with slower computers. And even if it is, it's not nice to consume a lot fo power and have loud fans just for rendering with Neovide.

I don't think storing pre-computed textures for each line is the answer either, the work still needs to be done, and I see no reason why this would have to be done. 

The old snapshot based approach is not very good either, it uses a lot of GPU memory, and it complicates a lot of the code.

So I will try to figure out what Skia is doing and see if there's anything to be done. But any ideas are welcome. The profiling I have done does not really help. I also have some very experimental internal profiling of Skia which I used before, but it turned out that is not very useful either.

Reply:
@fredizzimo this might sound insane but if skia is a problem is there an alternative to skia that you think would be a better fit, or is it faster to keep skia and optimize around it ?

Reply:
@A-Lamia, that's something I have been thinking about multiple times, I have spent so much time fighting with Skia, not just this time. It's just that it finally cooked over for me, when I though that I'm very close to have the rendering changes ready for a PR. It's also not just that it's a blackbox, the documentation is one of the worst I have seen for a software project that big. But on the other hand Skia is written by google for Google using it for Chrome, optimized for their usage patterns and so on.

Glfw is one framework that I have been considering https://www.glfw.org/, with rust support here https://docs.rs/glfw/latest/glfw/. That would also let us get rid of another of our painpoints winit, since it also includes window, keyboard and mouse handling. The problem is that it is only opengl, and from my experience I have not managed to get Windowed opengl smooth in Windows. Even examples by Nvidia are a bit hit and miss, it seems if you load the GPU enough, then it can be smooth, but otherwise there will be some choppyness.

I have not looked deeply at other options, but sdl might be possible https://www.libsdl.org/, that would probably also let us get rid of winit

Even using straight opengl and direct 3d would not be too hard. But all the animation effects would need to be re-done, but could probably be done completly as shaders, give much greater possibilities.

The thing is that all of this is a lot of effort, and still some uncertainy about the results.

Reply:
@fredizzimo glfw seems to support vulkan would that be on par with direct 3d ?

Reply:
I'm not sure about Vulkan, it's rather new and we might also run into issues with no support for older hardware. I'm also not sure if that would resolve the problems on Windows.

But I think I found a plan now. There seems to be a lot of overhead for each `draw_picture` it creates new command buffers, and flushes a lot of state. So the plan would be to instead do some preprocessing so, I can run a single `draw_vertices` for the backgrounds, and also a single skia call for all the glyps on the screen. And also some preprocesing for all the underlines.

Reply:
awesome, let us know how it goes! :)

Reply:
Funnily enough this project was originally based on sdl2, but we switched to winit + openGL after it gave us quite a few problems. The rust bindings might be more mature now though

Reply:
The thing I really don't understand where this performance regression has come frome. 

I finally got Skia 0.52 working again, after fixing its source code a bit. It stopped compiling from source after I updated Visual Studio. But the problem remains. (There's no binary package available for 0.52 and the the d3d/opengl combination. But there is for 0.54 and onwards, so I updated to that version in my rendering branch).

It was working fine before and super smooth, with almost the same drawing function as I have now in this PR. But now it's running 30 FPS when I open telescope, so even slower than the opengl version.

So I guess I should see if I can find some older versions either in some backup branch that I might have or in the git reflog, if that gives some clue.

I know that Skia should be able to handle this, but the problem is that the almost non-existent documentation and how abstracted it is from the actual hardware rendering that it does in the end. And the source code is also very complex and hard to follow with a lot of strange callbacks.

Reply:
Offtop: wgpu (https://wgpu.rs/) also supports vulkan and opengl, and has pretty good rust-native API. Not as easy as raw opengl (the pipeline initialization part), but more convenient than raw vulkan. Some examples can be seen here: https://sotrh.github.io/learn-wgpu, https://github.com/gfx-rs/wgpu/tree/master/wgpu/examples

Reply:
It turns out that setting a blend mode for `draw_picture`, like `Some(&foreground_paint)` is a massive slowdown. Even if it's just `BlendMode::Src` 

But commenting the background drawing out and and drawing the foreground without a blend mode gives black windows with only the bottom line visible. That seems to indicate that there's a massive overdraw, and that I need to set a clip region. I thought that the picture was only the size I created it with, but apparently not... 

On the other hand, that's not the same as `BlendMode::Src`, which correctly draws each line. So I have no idea what's going on actually. Like I said Skia is a big black box. 

I had a similar experience when trying to fix the transparency for my second pass of this PR, the blend modes and everything seemed to interact in a very strange way. And add to the fact that skia uses premultiplied alpha for all the render target surfaces, at least I think so, although I don't see anything in the official documentation. The chromium doucmentation is a little bit more explicit though https://www.chromium.org/developers/design-documents/graphics-and-skia/. I just eneded up doing trial and error for several hours until I found some magic combination that worked.

Anyway, I will continue this investigation another day.

Reply:
@CertainLach wgpu definitely looks interesting. I will take a closer look at it at some point.

Reply:
I've done some playing with wgpu and did some initial ports of zed's shader to it's system. I would LOVE if we could swap to it, but just havent put in the time to really go and do it.

We depend on skia for the text rasterization, font fallback, and rendering you've been discussing above. Text rasterization can be done pretty simply using swash, but font fallback is a much more complicated problem.

If we can address font fallback, I see no reason why we couldn't spend some effort swapping over pretty quickly.

Reply:
For the information. I stated experimenting a bit with wgpu here https://github.com/fredizzimo/neovide/tree/fsundvik/wgpu

It currently renders the background, with the correct colors, so you can see the status bar for example. Obviously no text yet, and there's som problems with scaling, the line height is not as tall as it should be.

For the fonts I was planning to keep our current swash shaper (at least for now, although I think we should switch at some pint), and use this https://github.com/alacritty/crossfont as the rasterizer and it looks like it handles the fallbaks as well. Wezterm has a nice atlas implementation https://github.com/wez/wezterm/blob/3666303c7b26c6c966b3f136dbb954686d334cc3/window/src/bitmaps/atlas.rs for caching of the actual rastered glyphs.

Another option is to use almost everyting from Wezterm here https://github.com/wez/wezterm/tree/3666303c7b26c6c966b3f136dbb954686d334cc3/wezterm-font. But I'm not so sure that using freetype on all platforms is a good idea.

I won't be spending a lot of time on this though, so if it's not working out smoothly, someone else can take over, or we can just keep on using Skia.

NOTE: I had to fight a lot with lifetime issues with wgpu, it really wants you to architect everything in a very specific way. But other than that everything has been smooth and quick to get up running so far. On the other hand, I'm still not that confident in writing rust, I really need to read some proper book that goes a bit deeper than "The Book".

Reply:
I'm curious about your desire to move away from swash. Can you elaborate?

I've also been working on a wgpu expirement of my own but using rust-gpu for the shader. Idk if this is a viable cross platform option but it has some really nice side effects like defining the instance types in rust and using them both shader and cpu side. I haven't gotten as far as you, but I've been documenting some of the progress here https://kaylees.dev/trio/maple/day5-quad-shader/

Reply:
Regarding swash, mostly because Harfbuzz is the standard almost everywhere and https://github.com/RazrFalcon/rustybuzz is available for Rust, so it would guarantee that the same font is shaped in the same way across all the applications a single user uses. 

We also have some issues where it's clear that we are rendering certain fonts in the wrong way and different than some terminal applications we are compared to. I'm not sure if that's a shaping issue, rasterizing issue, or that we don't use the metrics we get out of the shaper correctly though.

For the shader I would rather implement hot reload, so you can modify the shader on the fly, like shadertoy for example https://webglfundamentals.org/webgl/lessons/webgl-shadertoy.html. But maybe that model is compatible with rust-gpu as well, I will check it out later, and also read your blog.

And a small update on my progress, I have now replaced the skia font loading by font-kit in my branch. But there's at least one drawback, the system font fallback is currently Windows only, in font-kit. Our regular fallbacks should work though, it's just the system part that is missing. It should not be too hard to add support for the other platforms and send a pull request for them, especially since we have both the cross-font and Skia as examples, and it's just a matter of calling the correct system function on each platform. 

So, my next step is actual rasterization and rendering of the fonts. I don't think I will use the atlas from wezterm as I first planned, since our is much more simple with just single or double width characters.

Reply:
We used font-kit in the past. I think at some point we swapped to just skia and swash to reduce dependencies. I also recall fontkit's glyph properties being strange but I don't remember the details.

Reply:
As for rustybuzz, the swap to swash was about performance. For whatever reason the shaping with rustybuzz was pretty slow. We have made some upstream issues in swash about font problems and at least a couple of times gotten them fixed. I dont feel super strongly either way though.

Reply:
Am i insane or did smooth scroll stop working ?